### PR TITLE
Trigger deleting metadata from DB after carboniting has been completed [BA-6051]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: hybridCarboniteMetadataService/src/test/scala/cromwell/services/metadata/hybridcarbonite/CarboniteWorkerActorSpec.scala [hunk 1]
```diff
@@ -67,6 +68,14 @@ class CarboniteWorkerActorSpec extends TestKitSuite("CarboniteWorkerActorSpec")
...
 
       carboniteFreezerActor.send(carboniteWorkerActor, CarboniteWorkflowComplete(WorkflowId.fromString(workflowToCarbonite), Archived))
+
+      serviceRegistryActor.fishForSpecificMessage(10.second) {
+        case DeleteMetadataAction(_, _, _) => true
+      }
+
+      EventFilter.info(message = s"Completed deleting metadata from database for carbonited workflow: $workflowToCarbonite", occurrences = 1) intercept {
+        deleteMetadataActor.send(carboniteWorkerActor, DeleteMetadataSuccessfulResponse(WorkflowId.fromString(workflowToCarbonite)))
+      }
```
Comment (mcovarr) [inline hybridCarboniteMetadataService/src/test/scala/cromwell/services/metadata/hybridcarbonite/CarboniteWorkerActorSpec.scala]: how about a test for the unhappy path as well?

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarboniteWorkerActor.scala [hunk 2]
```diff
@@ -73,10 +73,22 @@ class CarboniteWorkerActor(carboniterConfig: HybridCarboniteConfig,
       // after completion of carbonite process reset the backoff
       log.info(s"Carboniting complete for workflow ${c.workflowId}")
 
+      if (c.result == Archived) {
```
Comment (mcovarr) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarboniteWorkerActor.scala]: could this be moved into the `case Archived` handling on line 68?

File: services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  implicit val ec = context.dispatcher
+
+  override def receive: Receive = {
+    case action@DeleteMetadataAction(workflowId, replyTo, maxAttempts) =>
+      val deleteFromDbFuture = deleteNonLabelMetadataEntriesForWorkflow(workflowId)
+      deleteFromDbFuture onComplete {
+        case Success(_) => replyTo ! DeleteMetadataSuccessfulResponse(workflowId)
+        case Failure(ex) if maxAttempts > 0 =>
+          val remainingAttempts = action.maxAttempts - 1
+          log.error(s"Cannot delete metadata. Remaining number of attempts: $maxAttempts. Error message: ${ex.getMessage}", ex)
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala]: not sure you need to put `ex.getMessage` in the string if you're also passing `ex` to `log.error`

File: services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  with MetadataServicesStore {
+
+  implicit val ec = context.dispatcher
+
+  override def receive: Receive = {
+    case action@DeleteMetadataAction(workflowId, replyTo, maxAttempts) =>
+      val deleteFromDbFuture = deleteNonLabelMetadataEntriesForWorkflow(workflowId)
+      deleteFromDbFuture onComplete {
+        case Success(_) => replyTo ! DeleteMetadataSuccessfulResponse(workflowId)
+        case Failure(ex) if maxAttempts > 0 =>
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala]: Won't this make up to two attempts if `maxAttempts` is 1?  Perhaps `maxRetries` would be more accurate.

File: services/src/main/scala/cromwell/services/metadata/impl/MetadataServiceActor.scala [hunk 5]
```diff
@@ -58,6 +58,7 @@ case class MetadataServiceActor(serviceConfig: Config, globalConfig: Config, ser
   val dbFlushRate = serviceConfig.getOrElse("db-flush-rate", 5.seconds)
   val dbBatchSize = serviceConfig.getOrElse("db-batch-size", 200)
   val writeActor = context.actorOf(WriteMetadataActor.props(dbBatchSize, dbFlushRate, serviceRegistryActor, LoadConfig.MetadataWriteThreshold), "WriteMetadataActor")
+  val deleteActor = context.actorOf(Props(new DeleteMetadataActor), "DeleteMetadataActor")
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/MetadataServiceActor.scala]: The `props` method in the companion object is a pretty standard pattern in Cromwell (examples above) though I'm not 💯 why

File: services/src/main/scala/cromwell/services/metadata/MetadataService.scala [hunk 6]
```diff
@@ -82,12 +82,14 @@ object MetadataService {
...
-  val MaximumMetadataWriteAttempts = 10
-  final case class PutMetadataAction(events: Iterable[MetadataEvent], maxAttempts: Int = MaximumMetadataWriteAttempts) extends MetadataWriteAction
-  final case class PutMetadataActionAndRespond(events: Iterable[MetadataEvent], replyTo: ActorRef, maxAttempts: Int = MaximumMetadataWriteAttempts) extends MetadataWriteAction
+  val MaximumMetadataActionAttempts = 10
+  final case class PutMetadataAction(events: Iterable[MetadataEvent], maxAttempts: Int = MaximumMetadataActionAttempts) extends MetadataWriteAction
+  final case class PutMetadataActionAndRespond(events: Iterable[MetadataEvent], replyTo: ActorRef, maxAttempts: Int = MaximumMetadataActionAttempts) extends MetadataWriteAction
 
   final case object ListenToMetadataWriteActor extends MetadataServiceAction with ListenToMessage
 
+  final case class DeleteMetadataAction(workflowId: WorkflowId, replyTo: ActorRef, maxAttempts: Int = MaximumMetadataActionAttempts) extends MetadataServiceAction
```
Comment (aednichols) [inline services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: I'm not convinced deletion benefits from retries... the failure modes for `deleteNonLabelMetadataEntriesForWorkflow` don't seem to be the sorts of things that will resolve themselves with time

Comment (grsterin) [reply services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: E.g., a random network glitch on attempt to reach the DB? Or DB being down for some amount of time and then being brought up?

Comment (mcovarr) [reply services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: we're already retrying puts...

Comment (aednichols) [reply services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: If we accept the premise that the DB connection can be flakey (I've never seen evidence of this, but could be true) then I was thinking we should push DB retries down into something like `cromwell.database.slick.SlickDatabase#runActionInternal`. It already has error handling & manages timeouts.

That said I agree `DeleteMetadataAction` matches an existing pattern

File: services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala (hunk 4 above)
Comment (mcovarr) [reply services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala]: it looks like `maxAttempts` is in the existing "puts" code and does actually represent the maximum number of attempts

File: hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarboniteWorkerActor.scala [hunk 7]
```diff
@@ -64,23 +64,39 @@ class CarboniteWorkerActor(carboniterConfig: HybridCarboniteConfig,
...
+      }
 
+      if (c.result != Archived) {
+        resetBackoffAndFindNextWorkflowForCarboniting()
+      }
+    case DeleteMetadataSuccessfulResponse(workflowId) =>
+      log.info(s"Completed deleting metadata from database for carbonited workflow: $workflowId")
+      resetBackoffAndFindNextWorkflowForCarboniting()
+    case DeleteMetadataFailedResponse(workflowId, reason) =>
+      log.error(s"All attempts to delete metadata from database for carbonited workflow $workflowId failed: {}", reason.getMessage)
```
Comment (mcovarr) [inline hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarboniteWorkerActor.scala]:
```suggestion
      log.error(reason, s"All attempts to delete metadata from database for carbonited workflow $workflowId failed")
```

File: services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  implicit val ec = context.dispatcher
+
+  override def receive: Receive = {
+    case action@DeleteMetadataAction(workflowId, replyTo, maxAttempts) =>
+      val deleteFromDbFuture = deleteNonLabelMetadataEntriesForWorkflow(workflowId)
+      deleteFromDbFuture onComplete {
+        case Success(_) => replyTo ! DeleteMetadataSuccessfulResponse(workflowId)
+        case Failure(ex) if maxAttempts > 1 =>
+          val remainingAttempts = action.maxAttempts - 1
+          log.error(s"Cannot delete metadata. Remaining number of attempts: $maxAttempts. Error message: {}", ex.getMessage)
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala]:
```suggestion
          log.error(ex, s"Cannot delete metadata. Remaining number of attempts: $maxAttempts.")
```

File: services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+          self ! action.copy(maxAttempts = remainingAttempts)
+        case Failure(ex) =>
+          log.error(s"Cannot delete metadata. Error message: {}", ex.getMessage)
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/impl/DeleteMetadataActor.scala]:
```suggestion
          log.error(ex, s"Cannot delete metadata.")
```

**Pull request merged**